File: 21_all_substrings.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 21: Generate all substrings of a string.
Prints each substring on its own line.
Warning: O(n^2) substrings.
*/

int main()
{
    string s;
    if (!getline(cin, s))
        return 0;
    int n = s.size();
    for (int i = 0; i < n; ++i)
        for (int len = 1; i + len <= n; ++len) cout << s.substr(i, len) << '\n';
    return 0;
}

// Time Complexity: O(n^2)
// Space Complexity: O(1) extra

File: 21_all_substrings.py
"""
Task 21: Generate all substrings of a string.
Prints each substring on its own line.
"""
import sys

if __name__ == '__main__':
    s = sys.stdin.readline().rstrip('\n')
    n = len(s)
    for i in range(n):
        for j in range(i+1, n+1):
            print(s[i:j])

# Time Complexity: O(n^2)
# Space Complexity: O(1) extra
